fix(android)(9_3_X): AudioRecorder recording/stopped property handling

jquick-axway opened the pull request:
**JIRA:**
https://jira.appcelerator.org/browse/TIMOB-28105

**Summary:**
- `AudioRecoder` changes:
  * Fixed bug where it reports the wrongs states via `recording` and `stopped` properties 
  * Now auto-stops when JS object is garbage collected or JS runtime is terminated.
  * Removed blocking calls from `start()` and `resume()` methods. _(Can block up to 10 sec.)_
  * Now checks if recording has actually started. _(Can't start if microphone is already in use.)_
  * Significantly optimized `stop()` method. _(No longer creates 2nd audio file.)_
- Added a new "build.post.install" build hook for Android and iOS.
- Added `AudioRecorder` unit tests for Android and iOS.
- Updated test suite's plugin to grant app permissions to iOS simulator and Android after app install.

**Test:**
1. Create a "Classic" app.
2. Add the below permission to the "tiapp.xml".
3. Use `AudioRecorderStatusTest.js` attached to [TIMOB-28105](https://jira.appcelerator.org/browse/TIMOB-28105) for the "app.js".
4. Build and run on Android.
5. Verify "Stopped" label is `true` and the other labels are `false`.
6. Tap on the "Start Recording" button.
7. Verify "Recording" label is `true` and the other labels are `false`.
8. Tap on the "Pause Recording" button.
9. Verify "Paused" label is `true` and the other labels are `false`.
10. Tap on the "Resume Recording" button.
11. Verify "Recording" label is `true` and the other labels are `false`.
12. Tap on the "Stop Recording" button.
13. Verify "Stopped" label is `true` and the other labels are `false`.

tiapp.xml
```xml
<ti:app>
	<android>
		<manifest>
			<uses-permission android:name="android.permission.RECORD_AUDIO"/>
		</manifest>
	</android>
</ti:app>
```

jquick-axway committed:
fix(android): AudioRecorder recording/stopped property handling

Fixes TIMOB-28105

jquick-axway committed:
feat: add "build.post.install" build hook

jquick-axway committed:
test: add AudioRecorder unit tests

build commented:
<!--
  1 failure:  Tests have failed...
  3 warning:  Commit <a href="h..., build/lib/test/te..., build/lib/test/te...
  3 messages
  1 markdown notices
  DangerID: danger-id-Danger;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>Tests have failed, see below for more information.</td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  Commit <a href="https://api.github.com/repos/appcelerator/titanium_mobile/commits/9b887eb24fb07b7fb8efb55e354f7de82ddfb853">9b887eb24fb07b7fb8efb55e354f7de82ddfb853</a> has a message "fix(android): AudioRecorder recording/stopped property handling

Fixes TIMOB-28105" giving 1 errors:
- subject must not be sentence-case, start-case, pascal-case, upper-case
  </td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  **build/lib/test/test.js#L443** - build/lib/test/test.js line 443 – 'stripped' is defined but never used. Allowed unused args must match /^_.+/u. (no-unused-vars)
  </td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  **build/lib/test/test.js#L775** - build/lib/test/test.js line 775 – Missing JSDoc parameter description for 'testResults'. (valid-jsdoc)
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-11996/1/artifact/dist/mobilesdk-9.3.0.v20200901165906-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>

  :rotating_light: This PR has one or more commits with warnings/errors for commit messages not matching our configuration. You may want to squash merge this PR and edit the message to match our conventions, or ask the original developer to modify their history.
  </td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>:x: 2 tests have failed
There are 2 tests failing and 734 skipped out of 9601 total tests.</td>
    </tr>
  </tbody>
</table>

### Tests:

<table><tr><th>Classname</th><th>Name</th><th>Time</th><th>Error</th></tr>
<tr><td>android.emulator.Titanium.Media.AudioRecorder</td><td>#start, #pause, #resume, #stop (11)</td><td>10.259</td><td><pre>Error: timeout of 5000ms exceeded
at Titanium.&lt;anonymous&gt; (/ti-mocha.js:6535:53723)</pre></td></tr>
<tr><td>ios.iphone.Titanium.UI.WebView</td><td>blacklistedURLs (14.0)</td><td>29.999</td><td><pre>Error: timeout of 30000ms exceeded
file:///Users/build/Library/Developer/CoreSimulator/Devices/B98CCEEE-6478-4929-BE6A-C6DB04CAF126/data/Containers/Bundle/Application/348B2BCB-7BB0-48A1-B552-3DF7A4A1281C/mocha.app/ti-mocha.js:4326:27</pre></td></tr>
</table>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 102464b45a94405c53b5a1732c49e07eaa06f8c9
</p>

jquick-axway committed:
test: update AudioRecorder test for disconnected microphone

jquick-axway committed:
refactor(android): audio recorder handling

- Now auto-stops recording when JS object is released or JS runtime has been terminated.
  * Needed to to free up microphone to be used by next audio recorder object.
- Modified AudioRecorder.start() to no-op if already started.
- Removed blocking reads from audio recorder. Now always reads microphone from worker thread.
- Now checks if audio recording was successfully started.

garymathews commented on android/modules/media/src/java/ti/modules/titanium/media/TiAudioRecorder.java:
```diff
@@ -18,12 +19,15 @@
 
 public class TiAudioRecorder
 {
+	private static final String TAG = "TiWindowProxy";
```
`TiAudioRecorder`

jquick-axway committed:
refactor(android): optimized audio recording to file

sgtcoolguy commented on tests/plugins/mocha.test.support/hooks/hook.js:
```diff
@@ -40,6 +40,25 @@ exports.init = (logger, config, cli) => {
+				await xcrun([ 'simctl', 'privacy', builder.simHandle.udid, 'grant', 'all', builder.tiapp.id ]);
+
+				// Re-launch app in case granting permissions forced-quit it.
+				await xcrun([ 'simctl', 'launch', builder.simHandle.udid, builder.tiapp.id ]);
```
I assume this is because of the note above where we're not actually firing post-install, pre-launch - we're firing post-launch?

jquick-axway replied:
Exactly right. We'd be able to remove this line if we could fire the post install event before the launch.

jquick-axway commented on android/modules/media/src/java/ti/modules/titanium/media/TiAudioRecorder.java:
```diff
@@ -38,134 +44,151 @@
-			paused = true;
-			audioRecord.stop();
+		if (this.audioRecord != null) {
+			this.audioRecord.stop();
```
Yes, we need the `null` checks. We set the `audioRecord` member variable to `null` when the `stopRecording()` method gets called. We create a new instance when the `startRecording()` method gets called.

garymathews replied:
Although `this.audioRecord.stop()` can be called multiple times, it might be good to do this:
```Java
if (!this.paused && this.audioRecord != null) {
```

jquick-axway replied:
An even better solution would be to ditch those 3 boolean member variables (which I agree is silly) and fetch the state from the audio recorder object itself similar to the below. Although I've already confirmed that there is no harm in calling the native `AudioRecord.start()` if already started and `AudioRecord.stop()` if already stopped.
```java
if (this.audioRecord == null) {
	// Stopped.
} else if (this.audioRecord.getRecordingState() == AudioRecord.RECORDSTATE_RECORDING) {
	// Recording.
} else {
	// Paused.
}
```

jquick-axway commented on iphone/cli/hooks/run.js:
```diff
@@ -90,8 +90,11 @@ exports.init = function (logger, config, cli) {
 				.on('app-started', function () {
-					finished && finished();
-					finished = null;
+					// TODO: Add "installed" event to "ioslib" module for simulators and emit below event from there.
```
Yes, I thought about making changes to ioslib, but I saw that major changes are being made to it on master. And I have no idea what's going on with the 2_X and 3_X branches. Are they used by the daemon and Appc Studio?

jquick-axway committed:
refactor(android): audio recording state handling

garymathews reviewed (approved):
CR: **PASS**

jquick-axway committed:
test: increased AudioRecorder timeout

ssekhri commented:
FR Passed.
Verified on:
Mac OS: 10.15.4
SDK: 9.3.0.v20201005141423
Appc CLI: 8.1.1
JDK: 11.0.4
Node: 10.17.0
Studio: 6.0.0.202005141803
Device: Pixel 3(v11.0) device, Pixel 3 XL(v11.0) emulator

build commented:
The backport to `master` failed:
```
The process 'git' failed with exit code 128
```
Check [the run](https://github.com/appcelerator/titanium_mobile/actions/runs/308815652) for full details
To backport manually, run these commands in your terminal:
```bash
# Fetch latest updates from GitHub
git fetch
# Check out the target branch
git checkout master
# Make sure it's up to date
git pull
# Check out your branch
git checkout -b backport-11996-to-master
# Apply the commits from the PR
curl -s https://github.com/appcelerator/titanium_mobile/commit/aa2d1ba047198f0624c2769a18dee70c00e5cb72.patch | git am -3 --ignore-whitespace
curl -s https://github.com/appcelerator/titanium_mobile/commit/22e8cc0a1e23487f82f9a039899214640834eaa0.patch | git am -3 --ignore-whitespace
curl -s https://github.com/appcelerator/titanium_mobile/commit/29277557f1fe68b49b39595f538ffe2a3629cad7.patch | git am -3 --ignore-whitespace
curl -s https://github.com/appcelerator/titanium_mobile/commit/8335e27ec6db2d3758bd637121973b430166ce94.patch | git am -3 --ignore-whitespace
curl -s https://github.com/appcelerator/titanium_mobile/commit/6c3869af9044ff9a107a4b8c692dbaa3cbcf5e57.patch | git am -3 --ignore-whitespace
curl -s https://github.com/appcelerator/titanium_mobile/commit/68a1057a2b897ccac65045fa1b655a7570376d9f.patch | git am -3 --ignore-whitespace
curl -s https://github.com/appcelerator/titanium_mobile/commit/487c403502fad18c7e6e531c56ce6c9f74dea796.patch | git am -3 --ignore-whitespace
curl -s https://github.com/appcelerator/titanium_mobile/commit/397cbb47ffde0b6a7e2ff97b725e0b5ecefbcf62.patch | git am -3 --ignore-whitespace
# Push it to GitHub
git push --set-upstream origin backport-11996-to-master
```
Then, create a pull request where the `base` branch is `master` and the `compare`/`head` branch is `backport-11996-to-master`.
